### PR TITLE
PHOENIX-4010 Hash Join cache may not be send to all regionservers whe…

### PR DESCRIPTION
**Pull request by ankitsinghal**
…n we have stale HBase meta cache

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/HashJoinIT.java [hunk 1]
```diff
@@ -3453,4 +3496,27 @@ public void testJoinWithSetMaxRows() throws Exception {
             conn.close();
         }
     }
+    
+    public static class InvalidateHashCacheRandomly extends SimpleRegionObserver {
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/HashJoinIT.java]: These join tests run on the same cluster as all other tests derived from ParallelStatsDisabledIT. Rather than having a random test failure, let's move this to a dedicated test derived from BaseUniqueNamesOwnClusterIT and have it consistently fail.

File: phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java [hunk 2]
```diff
@@ -394,4 +420,82 @@ public static String idToString(byte[] uuid) {
...
+                    });
+        } catch (Throwable t) {
+            throw new Exception(t);
+        }
+        if (results != null && results.size() == 1) { return results.values().iterator().next().getReturn(); }
+        return false;
+
+    }
+    
+    public static ServerCache getCacheForId(List<ServerCache> caches, Long cacheId) {
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java]: How about using a Map<ImmutableBytesPtr,ServerCache> instead of a List and switching id to ImmutableBytesPtr? Then you can get rid of this static method.

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 3]
```diff
@@ -163,8 +185,38 @@ public Tuple next() throws SQLException {
...
+                                    "Retrying when Hash Join cache is not found on the server ,by sending the cache again");
+                            if (retry <= 0) { throw e1; }
+                            retry--;
+                            try {
+                                Long cacheId = ((HashJoinCacheNotFoundException)e1).getCacheId();
+                                if (!hashCacheClient.addHashCacheToServer(newScan.getStartRow(),
+                                        ServerCacheClient.getCacheForId(caches, cacheId), plan.getTableRef().getTable())) { throw e1; }
+                                
+                                try {
+                                    if(ScanUtil.isClientSideUpsertSelect(newScan)) {
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java]: Why is this necessary? Can we generalize this is a better way?

File: phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java [hunk 4]
```diff
@@ -303,73 +318,84 @@ public TaskExecutionMetricsHolder getTaskExecutionMetric() {
...
+            List<byte[]> keys = cache.getKeys();
+            ConnectionQueryServices services = connection.getQueryServices();
+            Throwable lastThrowable = null;
+            TableRef cacheUsingTableRef = cacheUsingTableRefMap.get(Bytes.mapKey(cacheId));
+            final PTable cacheUsingTable = cacheUsingTableRef.getTable();
+            byte[] tableName = cacheUsingTableRef.getTable().getPhysicalName().getBytes();
+            iterateOverTable = services.getTable(tableName);
+
+            List<HRegionLocation> locations = services.getAllTableRegions(tableName);
+            Set<HRegionLocation> remainingOnServers = new HashSet<HRegionLocation>(servers);
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java]: Instead of Set<HRegionLocation> can we change this to Set<ServerName>? It'd make it more clear IMHO

File: phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java [hunk 5]
```diff
@@ -303,73 +318,84 @@ public TaskExecutionMetricsHolder getTaskExecutionMetric() {
...
+             * through the current metadata boundaries and remove the cache once for each server that we originally sent
+             * to.
+             */
+            if (LOG.isDebugEnabled()) {
+                LOG.debug(addCustomAnnotations("Removing Cache " + cacheId + " from servers.", connection));
+            }
+            for (HRegionLocation entry : locations) {
+             // Call once per server
+                if (remainingOnServers.contains(entry)
+                        || (keys != null && ByteUtil.contains(keys,entry.getRegionInfo().getStartKey()))) {
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/cache/ServerCacheClient.java]: Why is this extra check required? Is this when the Hash Cache is sent over to the RS that doesn't have it? If so, can we document that?

File: phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java [hunk 6]
```diff
@@ -804,6 +804,7 @@ public Long getEstimatedBytesToScan() throws SQLException {
             ////////////////////////////////////////////////////////////////////
             // UPSERT SELECT run client-side
             /////////////////////////////////////////////////////////////////////
+            ScanUtil.setClientSideUpsertSelect(queryPlan.getContext().getScan());
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java]: This would be better to encapsulate in the StatementPlan if we even need it. Why do we need it?

File: phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java [hunk 7]
```diff
@@ -108,6 +126,10 @@ public TableResultIterator(MutationState mutationState, Scan scan, ScanMetricsHo
         this.scanIterator = UNINITIALIZED_SCANNER;
         this.renewLeaseThreshold = renewLeaseThreshold;
         this.scanGrouper = scanGrouper;
+        this.hashCacheClient = new HashCacheClient(plan.getContext().getConnection());
+        this.caches = caches;
+        this.retry=plan.getContext().getConnection().getQueryServices().getConfiguration()
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/iterate/TableResultIterator.java]: Can we use getProps() here instead of going through the Configuration object? We've found the synchronization done by Configuration to be a bottleneck in the past (and not necessary since these properties don't change).

File: phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java (hunk 6 above)
Comment (ankitsinghal) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java]: this change is actually a part of PHOENIX-3163, let me see if we can improve.

Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/compile/UpsertCompiler.java]: Let's do those changes in a separate patch, please.

**Comment by ankitsinghal** [comment]
@JamesRTaylor , can you please review, I have removed the dependency on PHOENIX-3163.

**Comment by ankitsinghal** [comment]
Ping @JamesRTaylor

File: phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java [hunk 8]
```diff
@@ -109,12 +110,16 @@
 
     /** Exposed for testing */
     public static final String SCANNER_OPENED_TRACE_INFO = "Scanner opened on server";
+    public static final String CLIENT_SIDE_UPSERT_SELECT = "_ClientSideUpsertSelect";
+    protected Configuration rawConf;
```
Comment (JamesRTaylor) [inline phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java]: What's the reason for this new rawConf member variable? Where is it used?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseJoinIT.java [hunk 9]
```diff
@@ -30,14 +30,19 @@
...
 import org.apache.phoenix.util.SchemaUtil;
 import org.apache.phoenix.util.StringUtil;
 import org.junit.Before;
+import org.junit.BeforeClass;
 
 import com.google.common.collect.ImmutableMap;
 import com.google.common.collect.Maps;
 
-public abstract class BaseJoinIT extends ParallelStatsDisabledIT {
+public abstract class BaseJoinIT extends BaseUniqueNamesOwnClusterIT {
```
Comment (JamesRTaylor) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/BaseJoinIT.java]: Can we keep BaseJoinIT derived from ParallelStatsDisabledIT? Otherwise, there's a lot more mini clusters created which increases our end-to-end test time run more and more.

If the reason is because you're adding InvalidateHashCacheRandomly, just add it instead dynamically to the table created in your test. To prevent flapping, you need to poll until you the descriptor matches what you've changed. See PartialIndexRebuilderIT.addWriteFailingCoprocessor() for an example you can generalize and/or copy/paste.

**Comment by JamesRTaylor** [comment]
Thanks for the ping, @ankitsinghal. The code looks good, but if we can fix up the tests with a couple of tweaks, we'll be good to go after rebasing.

**Comment by ankitsinghal** [comment]
Thanks @JamesRTaylor , have taken care the review comments in latest commit. Let me know if it is fine now

**Comment by JamesRTaylor** [comment]
+1

**Comment by stoty** [comment]
Already merged.
